src/thor/triplegbuilder.cc

```diff
@@ -91,93 +91,190 @@ void AssignAdmins(const AttributesController& controller,
+    // cutting for now
+    const auto& traffic_speed = tile->trafficspeed(edge);
+    if (traffic_speed.breakpoint1 > 0) {
+      cuts.emplace_back(cut_t{traffic_speed.breakpoint1 / 255.0, speed,
+                              static_cast<std::uint8_t>(traffic_speed.congestion1)});
+      if (traffic_speed.breakpoint2 > 0) {
+        cuts.emplace_back(cut_t{traffic_speed.breakpoint2 / 255.0, speed,
+                                static_cast<std::uint8_t>(traffic_speed.congestion2)});
+        if (traffic_speed.speed3 != UNKNOWN_TRAFFIC_SPEED_RAW) {
+          cuts.emplace_back(cut_t{1, speed, static_cast<std::uint8_t>(traffic_speed.congestion3)});
         }
       }
-
-      // TODO Inject and cut for incidents here
-    }
-    // Cap the end so that we always have something to use
-    if (speeds.empty() || std::get<0>(speeds.back()) < tgt_pct) {
-      speeds.emplace_back(tgt_pct, speed);
     }
+  }
 
-    // Set the shape attributes
-    double distance_total_pct = src_pct;
-    auto speed_itr = std::find_if(speeds.cbegin(), speeds.cend(),
-                                  [distance_total_pct](const decltype(speeds)::value_type& s) {
-                                    return distance_total_pct <= std::get<0>(s);
-                                  });
-    assert(speed_itr != speeds.cend());
-    for (auto i = shape_begin + 1; i < shape.size(); ++i) {
-      // If there is a change in speed here we need to make a new shape point and continue from
-      // there
-      double distance = shape[i].Distance(shape[i - 1]); // meters
-      double distance_pct = distance / edge->length();
-      double next_total = distance_total_pct + distance_pct;
-      size_t shift = 0;
-      if (next_total > std::get<0>(*speed_itr) && std::next(speed_itr) != speeds.cend()) {
-        // Calculate where the cut point should be between these two existing shape points
-        auto coef =
-            (std::get<0>(*speed_itr) - distance_total_pct) / (next_total - distance_total_pct);
-        auto point = shape[i - 1].PointAlongSegment(shape[i], coef);
-        shape.insert(shape.begin() + i, point);
-        next_total = std::get<0>(*speed_itr);
-        distance *= coef;
-        shift = 1;
-      }
-      distance_total_pct = next_total;
-      double time = distance / std::get<1>(*speed_itr); // seconds
+  // Cap the end so that we always have something to use
+  if (cuts.empty() || cuts.back().percent_along < tgt_pct) {
+    cuts.emplace_back(cut_t{tgt_pct, speed, UNKNOWN_CONGESTION_VAL, {}});
+  }
 
-      // Set shape attributes time per shape point if requested
-      if (controller.attributes.at(kShapeAttributesTime)) {
-        // convert time to milliseconds and then round to an integer
-        trip_path.mutable_shape_attributes()->add_time((time * kMillisecondPerSec) + 0.5);
+  // sort the start and ends of the incidents along this edge
+  for (const auto& i : incidents) {
```

Comment:
this bit is new. here we add more "cut locations" based on where incident occur on the edge, if any. have a read over it, its well commented

Reply:
Could `i` be named `incident`? Makes it so much easier to read.

Reply:
sure thing!